File: CollectionLayout-OC/CollectionLayout/ControllerGroup/CardReelViewController.m
//
//  CardReelViewController.m
//  CollectionLayout
//
//  Created by 苏沫离 on 2020/8/12.
//  Copyright © 2020 苏沫离. All rights reserved.
//

#define CellIdentifer @"CollectionCardCell"

#import "CardReelViewController.h"
#import "DataModel.h"
#import "CollectionCardCell.h"
#import "YLCollectionReelLayout.h"

@interface CardReelViewController ()
<UICollectionViewDelegate,UICollectionViewDataSource>

@property (nonatomic ,strong) UICollectionView *collectionView;

@end

@implementation CardReelViewController

- (void)viewDidLoad{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    self.view.backgroundColor = [[UIColor colorWithPatternImage:[UIImage imageNamed:@"CardBack"]] colorWithAlphaComponent:0.9];
    [self.view addSubview:self.collectionView];
}

- (void)viewWillLayoutSubviews{
    [super viewWillLayoutSubviews];
}

#pragma mark - UICollectionViewDelegate

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return DataModel.shareDemoData.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    CollectionCardCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:CellIdentifer forIndexPath:indexPath];
    DataModel *model = DataModel.shareDemoData[indexPath.row];
    model.index = indexPath.row;
    cell.model = model;
    return cell;
}

#pragma mark - getter and setter

- (UICollectionView *)collectionView{
    if (_collectionView == nil) {
        YLCollectionReelLayout *layout = [[YLCollectionReelLayout alloc] init];
        layout.itemSize = CGSizeMake(200, 240);
        layout.radius = 500;
        
        UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.view.bounds),  CGRectGetWidth(self.view.bounds)) collectionViewLayout:layout];
        collectionView.dataSource = self;
        collectionView.delegate = self;
        collectionView.showsVerticalScrollIndicator = NO;
        collectionView.showsHorizontalScrollIndicator = NO;
        [collectionView registerClass:CollectionCardCell.class forCellWithReuseIdentifier:CellIdentifer];
        collectionView.backgroundColor = UIColor.clearColor;
        _collectionView = collectionView;
    }
    return _collectionView;
}
@end

File: CollectionLayout-Swift/CollectionLayout-Swift/ControllerGroup/CardReelViewController.swift
//
//  CardReelViewController.swift
//  CollectionLayout-Swift
//
//  Created by 苏沫离 on 2020/9/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//
// 转轮效果

import UIKit

class CardReelViewController: UIViewController,UICollectionViewDelegate,
                              UICollectionViewDataSource{
   ///懒加载
   lazy var collectionView : UICollectionView! = {
        let flowLayout = YLCollectionReelLayout()
        flowLayout.itemSize = CGSize(width: 200, height: 240)
        flowLayout.radius = 500

        let collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: flowLayout)
        collectionView.backgroundColor = UIColor.clear
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.delegate = self
        collectionView.dataSource = self
        collectionView.register(CollectionCardCell.self, forCellWithReuseIdentifier: "CollectionCardCell")
        return collectionView
   }()

   
   // MARK: - life cycyle
   override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor(patternImage: UIImage(named: "CardBack")!).withAlphaComponent(0.9)
        self.view.addSubview(collectionView)
   }
   
   // MARK: - UICollectionViewDelegate
   func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
       return DataModel.shareDemoData().count
   }
   
   func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
       if kind == UICollectionView.elementKindSectionHeader {
           let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HeaderView", for: indexPath)
           return headerView
       }
       let footerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footerView", for: indexPath)
       return footerView
   }
   
   func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell : CollectionCardCell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionCardCell", for: indexPath) as! CollectionCardCell
        let model = DataModel.shareDemoData()[indexPath.row]
        model.index = indexPath.row
        cell.model = model
        return cell
   }
}


